certbot-ci/certbot_integration_tests/utils/misc.py

```diff
@@ -30,16 +40,142 @@ def check_until_timeout(url):
+
+    cert = crypto.load_certificate(crypto.FILETYPE_PEM, data.encode('utf-8'))
+    return crypto.dump_certificate(crypto.FILETYPE_TEXT, cert).decode('utf-8')
+
+
+class GraceFullTCPServer(socketserver.TCPServer):
+    allow_reuse_address = True
+
+
 @contextlib.contextmanager
-def execute_in_given_cwd(cwd):
-    """
-    Context manager that will execute any command in the given cwd after entering context,
-    and restore current cwd when context is destroyed.
-    :param str cwd: the path to use as the temporary current workspace for python execution
-    """
+def create_tcp_server(port):
     current_cwd = os.getcwd()
+    webroot = tempfile.mkdtemp()
+
+    def run():
+        GraceFullTCPServer(('', port), SimpleHTTPServer.SimpleHTTPRequestHandler).serve_forever()
+
+    process = multiprocessing.Process(target=run)
+
     try:
-        os.chdir(cwd)
-        yield
+        try:
+            os.chdir(webroot)
+            process.start()
+        finally:
+            os.chdir(current_cwd)
+
+        check_until_timeout('http://localhost:{0}/'.format(port))
+
+        yield webroot
     finally:
-        os.chdir(current_cwd)
+        try:
+            if process.is_alive():
+                process.terminate()
+                process.join()  # Block until process is effectively terminated
+        finally:
+            shutil.rmtree(webroot)
+
+
+def list_renewal_hooks_dirs(config_dir):
+    renewal_hooks_root = os.path.join(config_dir, 'renewal-hooks')
+    return [os.path.join(renewal_hooks_root, item) for item in ['pre', 'deploy', 'post']]
+
+
+def generate_test_file_hooks(config_dir, hook_probe):
+    if sys.platform == 'win32':
+        extension = 'bat'
+    else:
+        extension = 'sh'
+
+    renewal_hooks_dirs = list_renewal_hooks_dirs(config_dir)
+
+    for hook_dir in renewal_hooks_dirs:
+        try:
+            os.makedirs(hook_dir)
+        except OSError as error:
+            if error.errno != errno.EEXIST:
+                raise
+        hook_path = os.path.join(hook_dir, 'hook.{0}'.format(extension))
+        if extension == 'sh':
+            data = '''\
+#!/bin/bash -xe
+if [ "$0" == "{0}" ]; then
+    if [ -z "$RENEWED_DOMAINS" -o -z "$RENEWED_LINEAGE" ]; then
+        echo "Environment variables not properly set!" >&2
+        exit 1
+    fi
+fi
+echo $(basename $(dirname "$0")) >> "{1}"\
+'''.format(hook_path, hook_probe)
+        else:
+            data = '''\
+
+'''
+        with open(hook_path, 'w') as file:
+            file.write(data)
+        os.chmod(hook_path, os.stat(hook_path).st_mode | stat.S_IEXEC)
+
+
+def manual_http_hooks(http_server_root):
+    auth = (
+        '{0} -c "import os; '
+        "challenge_dir = os.path.join('{1}', '.well-known/acme-challenge'); "
+        'os.makedirs(challenge_dir); '
+        "challenge_file = os.path.join(challenge_dir, os.environ.get('CERTBOT_TOKEN')); "
+        "open(challenge_file, 'w').write(os.environ.get('CERTBOT_VALIDATION')); "
+        '"'
```

Comment:
Previously the script waited for the server to pick up the new file. How do we know it's picked it up in this version?

Reply:
This would imply a loop with a conditional break, and this can only be written with indentation and carriage returns. Indentation and carriage returns are not possible in a on-the-fly python code execution like with `python -c`. And this needs to be a on-the-fly python code execution because it is passed to the `--manual-auth-hook` flag of certbot in a shell.

As I observed, the tests had never failed because of the absence of this check in certbot-ci. Also to consider that the ACME server is making several attempts on his side to retrieve the challenge, so it already covers the situation of the challenge not being immediately served by the HTTP server.

Alternative solution here would be to write the content of the script in a temporary location, then pass `python /temp/path/to/script.py` to `--manual-auth-hook`.

However, since the situation is already covered and never failed with current implementation, I would like to avoid to add unnecessary complexity here.

Reply:
So first of all, I think in general if you're going to make a change like this during what's mostly a strict code translation, it's good to note that both for the reviewer and for posterity. Makes me wonder if there's other similar places in the code that I've missed?

Second, `while` loops with simple statements can be written in one line in Python.

Reply:
It did not say a loop. I said a loop with conditional break:
```
while predicate1:
  if predicate2:
    break
  do_logic()
```

This is not possible in one line.

Reply:
For your first remark, I made a lot of efforts to keep things as close as possible to the bash script in order to ease the review, even if it makes some weird python code (but planning to make a refactor afterward). Some modifications like this are more in the scope of an error, by forgetting to put a proper comment about it.

It is however the goal of a PR review, to track theses errors or get a clarification, and the sense to the effort I put also to give the shortest PRs I can with less than 400 LOC.

I would like also to remind that this code, as designed as a test code to raise for any deviation to the specs, is running successfully and consistently since 4 months, with an update from master each week, and with an equivalent code coverage compared to the original integration tests. So even if I made some slight modifications accidentally not documented in the implementation, they will be minor without effect to the capability of certbot-ci to catch bugs.

Reply:
Why does it need a second predicate? What's wrong with a `while requests.get(url).status_code != 200` or whatever?

Reply:
Oh, I remember now why I removed that loop in the first place. Because this is wrong: this kind of loop with a unique predicate will loop forever if we fall in a situation where the predicate will never change its state. Typical risk when the predicate represents in fact a call to an external component of unknown state, like an HTTP request.

A correct code is to have a second independent predicate whose you control the state, like a counter, to break the loop at some point. This implies two level of branching at least, as said before.

But anyway, concerning one level of branching loops, if a one-line loop can work when the body of the loop contain a unique simple statements in a python script, this is not the case from the command line. See failure of `python -c "import sys; if true: print sys.write('ok');"` for instance.

As the original loop has a flaw, as the behavior without the loop is consistently correct for serveral months, can we just end to write a comment about why there is not control of the state, why it is not a problem, and move on ?

Reply:
We could easily set a timeout here, and have it be of the form `while current_time < deadline and (requesting the url is still failing)`. That's a good improvement. But as a larger issue, I think it reduces clarity and increases code complexity to mix translation/refactoring and code improvements in the same PR. This PR is about translating the existing test to python, and we've had that check in that form for quite some time without issue.

Reply:
@ohemorange, I think you missed some points from my previous comment here.

Reply:
So, as I said, my point was not about if I should or not keep the old behavior to check the token file been correctly deployed, but if I can or cannot. For the should, yes obviously it is better if possible, even if I noted that the previous implementation was theoretically unsafe. For the can, having the python code passed as a command to the python interpreter forbid most of usages of indentation, required here for any `if/while/for` branches to keep the old behavior.

So as a general note, try to be careful during the review about the context of a conversation, take account in particular that non-native English speakers may use improperly some idiomatic expressions that gives an incorrect interpretation: maybe it is the case or not here, but I am willing to rephrase anything I said to be sure that it is fully understood.

Now for the specific issue here, I refactored to circumvent the initial issue of not been able to make some controls during the authentication script execution: now a python file is generated on-the-fly (deleted upon context exiting), and passed as an argument to the python interpreter to constitute an executable command for the certbot CLI.

By being a python file, every indentation is available, and so I implemented the condition on file availability you required. I also improved it, to avoid an infinite loop, by adding a timeout mechanism.

The command generated will be compatible both for Linux and Windows.